trulens_eval/trulens_eval/guardrails/langchain.py

```diff
@@ -0,0 +1,138 @@
+        
+        Args:
+            feedback (Feedback): use this feedback function to score each document.
+
+            threshold (float): and keep documents only if their feedback value is at least this threshold.
+
+        !!! example "Using TruLens guardrail context filters with Langchain"
+
+            ```python
+            from trulens_eval.guardrails.langchain import WithFeedbackFilterDocuments
+
+            # note: feedback function used for guardrail must only return a score, not also reasons
+            f_context_relevance_score = (
+                Feedback(provider.context_relevance)
+                .on_input()
+                .on(context)
+                .aggregate(np.mean)
+            )
+
+            filtered_retriever = WithFeedbackFilterDocuments.of_retriever(
+                    retriever=retriever,
+                    feedback=f_context_relevance_score,
+                    threshold=0.5
+                )
+
+            rag_chain = (
+                {"context": filtered_retriever | format_docs, "question": RunnablePassthrough()}
+                | prompt
+                | llm
+                | StrOutputParser()
+            )
+
+            tru_recorder = TruChain(rag_chain,
+                app_id='Chain1_ChatApplication_Filtered',
+                feedbacks=[f_answer_relevance, f_context_relevance, f_groundedness])
+
+            with tru_recorder as recording:
+                llm_response = rag_chain.invoke("What is Task Decomposition?")
+            ```  
+        """
+
+        super().__init__(
+            *args, feedback=feedback, threshold=threshold, **kwargs
+        )
+
+    # Signature must match
+    # langchain.schema.retriever.BaseRetriever._get_relevant_documents .
+    def _get_relevant_documents(self, query: str, *,
+                                run_manager) -> List[Document]:
+        """
+        An internal method to accomplish three tasks:
+
+        1. Get relevant documents.
+        2. Evaluate documents with a specified feedback function.
+        3. Filter out documents that do not meet the minimum threshold.
+
+        Args:
+            query: str - the query string to search for relevant documents.
+            
+            run_manager: RunManager - the run manager to handle document retrieval.
+
+        Returns:
+        - List[Document]: a list of filtered, relevant documents.
+        """
+        # Get relevant docs using super class:
+        docs = super()._get_relevant_documents(query, run_manager=run_manager)
+
+        # Evaluate the filter on each, in parallel.
+        ex = ThreadPoolExecutor(max_workers=max(1, len(docs)))
+
+        futures = list(
+            (
+                doc,
+                ex.submit(
+                    lambda doc, query: self.feedback(query, doc.page_content),
+                    doc,
+                    query
+                )
+            ) for doc in docs
+        )
+        wait([future for (_, future) in futures])
```

Comment:
iirc we can filter on demand instead of waiting for all feedbacks to finish
https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor-example